2022Swerve/src/main/java/frc/robot/subsystems/SimulationSubsystem.java

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+//This code belongs to Asher123456789
+
+
+
+package frc.robot.subsystems;
+
+import edu.wpi.first.hal.SimDouble;
+import edu.wpi.first.hal.simulation.SimDeviceDataJNI;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.ChassisSpeeds;
+import edu.wpi.first.math.kinematics.SwerveDriveOdometry;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj.smartdashboard.Field2d;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+
+public class SimulationSubsystem extends SubsystemBase {
+
+  private final Field2d m_field;
+  
+  private final SwerveSubsystem swerveSubsystem;
+  private final SwerveDriveOdometry m_odometry;
+  private int navXSim = SimDeviceDataJNI.getSimDeviceHandle("navX-Sensor[0]");
+  private double simYaw = 0;
+  /** Creates a new SimulationSubsystem. */
+  public SimulationSubsystem(SwerveSubsystem swerveSubsystem) {
+    this.swerveSubsystem = swerveSubsystem;
+
+    //This defines m_odometry with the DriveKinematics, Rotation of the swerve, and the robot's position
+    m_odometry = new SwerveDriveOdometry(Constants.DriveConstants.kDriveKinematics, swerveSubsystem.getRotation2d(), new Pose2d(0, 0, new Rotation2d()));
```

Comment:
Nothing we need to change here yet, but odometry should move over to the swerve subsystem in the future.

Reply:
Okay